# ¡Hola Natalia! 😊

Mi nombre es **Alejandro Castellanos** y hoy tengo el placer de ser el revisor de tu proyecto.

Voy a revisar todo tu código con detalle, buscando tanto los puntos fuertes como aquellos en los que podrías mejorar. Te dejaré comentarios a lo largo del notebook, destacando lo que has hecho bien y sugiriendo ajustes donde sea necesario. Si encuentro algún error, no te preocupes, te lo haré saber de forma clara y te daré información útil para que puedas corregirlo en la próxima iteración. Si en algún punto tienes comentarios, siéntete libre de dejarlos también.


Encontrarás mis comentarios específicos dentro de cajas verdes, amarillas o rojas, es muy importante que no muevas, modifiques o borres mis comentarios, con el fin de tener un seguimiento adecuado de tu proceso:


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class=“tocSkip”></a>
Si todo está perfecto.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class=“tocSkip”></a>
Si tu código está bien pero se puede mejorar o hay algún detalle que le hace falta.
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class=“tocSkip”></a>
Si de pronto hace falta algo o existe algún problema con tu código o conclusiones.
</div>

Puedes responderme de esta forma:
<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class=“tocSkip”></a>
</div>

A continuación te dejaré un comentario general con mi valoración del proyecto. **¡Mi objetivo es que sigas aprendiendo y mejorando con cada paso!**

---

<div class="alert alert-block alert-success">
<b>Comentario General del revisor (1ra Iteración)</b> <a class=“tocSkip”></a>

Natalia quiero felicitarte por el excelente trabajo que hiciste. Has realizado un análisis exploratorio de datos (EDA) apropiado, preparando la información de manera cuidadosa para el entrenamiento de los modelos predictivos, lo cual establece una base sólida para las etapas posteriores. Tu implementación de tus modelos de clasifiación ha sido acertada en combinación con técnicas de manejo de clases desbalanceadas, logrando alcanzar los objetivos establecidos en términos del F1-Score, lo que demuestra tu capacidad para ajustar modelos de clasificación efectivos.  
    
    
Recuerda que en el aprendizaje automático, la exploración de diferentes técnicas y enfoques es clave para desarrollar una comprensión más profunda y encontrar las soluciones más efectivas para cada problema específico. Acá te comparto técnicas adicionales de manejo de equilibrio de clases que te pueden resultar útiles en otros proyectos: [10 técnicas para resolver clases desequilibradas en ML](https://www.analyticsvidhya.com/articles/class-imbalance-in-machine-learning/)
    
Te deseo éxitos en tu próximo sprint 🚀
    
    
*Estado del Proyecto:* **Aprobado**

</div>

---

# Proyecto - Modelo para predecir salida de clientes de Beta Bank

# Paso 1. Descarga y preparación de los datos

In [1]:
#Importación inicial de librerias
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler


In [2]:
# Cargar los datos
data = pd.read_csv('/datasets/Churn.csv')

# Exploración inicial
print(data.head())  # Primeras filas
print(data.info())  # Información general
print(data.describe())  # Estadísticas descriptivas

# Verificar valores faltantes
print(data.isnull().sum())

   RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         790

<div class="alert alert-block alert-success">
<b>Comentario del revisor (1ra Iteración)</b> <a class=“tocSkip”></a>

Muy buen trabajo importando las librerías y los datos del proyecto. Adicionalmente usaste correctamente las funciones `info`, `describe`y `head`, esto te permite hacer una primera revisión de los datos, su estructura y contenido. Con esta información, podemos establecer una hoja de ruta para ajustar, modificar y analizar los datos de una manera adecuada.

</div>

In [3]:
# Eliminar columnas irrelevantes
data = data.drop(['RowNumber', 'CustomerId', 'Surname'], axis=1)

# Imputar valores nulos en la columna 'Tenure' con la mediana
median_tenure = data['Tenure'].median()
data['Tenure'].fillna(median_tenure, inplace=True)

# Verificar que no queden valores nulos
print(data.isnull().sum())

# Codificar variables categóricas
data = pd.get_dummies(data, columns=['Geography', 'Gender'], drop_first=True)

# Dividir en características (X) y objetivo (y)
X = data.drop('Exited', axis=1)
y = data['Exited']

# Dividir en conjuntos de entrenamiento, validación y prueba
X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size=0.3, random_state=12345)
X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=12345)

# Estandarizar características numéricas
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_val = scaler.transform(X_val)
X_test = scaler.transform(X_test)

CreditScore        0
Geography          0
Gender             0
Age                0
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          0
IsActiveMember     0
EstimatedSalary    0
Exited             0
dtype: int64


<div class="alert alert-block alert-success">
<b>Comentario del revisor (1ra Iteración)</b> <a class=“tocSkip”></a>

Buen trabajo manejando los datas faltantes de Tenure, esto es clave para evitar errores en el desarrollo de los modelos. Adicionalmente, eliminastes las columnas que no nos aportan al desarrollo de los modelos predictivos. Por otro lado, la codificación de las variables categóricas y el escalado de las valores numéricos le da mayor robustez a los modelos y mejora su rendimiento ,
    .

<div class="alert alert-block alert-warning">
<b>Comentario del revisor (1ra Iteración)</b> <a class=“tocSkip”></a>

Con respecto a la imputación de datos faltantes, si bien usar la mediana es una opción válida, hay otros métodos que pueden ser más adecuados y que pueden reducir la posibilidad de agregar sesgo a los datos, acá te comparto un artículo donde hablan sobre diferentes [métodos de imputación](https://www.kaggle.com/code/parulpandey/a-guide-to-handling-missing-values-in-python)

</div>
</div>

# Paso 2. Examinar el equilibrio de clases y modelo básico

In [4]:
# Contar las clases
print(y_train.value_counts())

# Porcentaje de cada clase
print(y_train.value_counts(normalize=True))

0    5589
1    1411
Name: Exited, dtype: int64
0    0.798429
1    0.201571
Name: Exited, dtype: float64


Clase 0 (No se fueron): 5,589 observaciones (79.84%).

Clase 1 (Se fueron): 1,411 observaciones (20.16%).

Esto confirma que hay un desequilibrio de clases, ya que la clase 0 es aproximadamente 4 veces más frecuente que la clase 1. 

<div class="alert alert-block alert-success">
<b>Comentario del revisor (1ra Iteración)</b> <a class=“tocSkip”></a>

Muy bien Natalia, has identificado que el dataset está desbalanceado. Esto es clave para definir el plan de trabajo

<div class="alert alert-block alert-warning">
<b>Comentario del revisor (1ra Iteración)</b> <a class=“tocSkip”></a>

Este tipo de análisis es conveniente hacerlo antes de hacer la segmetnación de los datos en entrenamiento, validación y prueba. Adicionalmente, si quieres que esta segmentación conserve la proporción de cada clase debes usar el parámetro `stratify`
    
    X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size=0.3, stratify=y, random_state=12345)

Esto asegura que la proporción de clases en `y_train` y `y_temp` sea similar a la de `y` original.
</div>    
    
</div>

In [5]:
#Modelo Básico

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score, roc_auc_score

# Entrenar modelo de regresión logística
model = LogisticRegression(random_state=12345, solver='liblinear')
model.fit(X_train, y_train)

# Predecir en el conjunto de validación
y_pred = model.predict(X_val)

# Calcular F1 y AUC-ROC
f1 = f1_score(y_val, y_pred)
auc_roc = roc_auc_score(y_val, model.predict_proba(X_val)[:, 1])

print(f'F1 Score: {f1:.3f}')
print(f'AUC-ROC: {auc_roc:.3f}')

F1 Score: 0.319
AUC-ROC: 0.776


Un valor del F1 Score (0.319) sugiere que el modelo tiene dificultades para predecir correctamente la clase positiva (clientes que se van). Esto es común en conjuntos de datos desequilibrados, donde la clase minoritaria (en este caso, Exited = 1) no está bien representada.

AUC-ROC (0.776): El AUC-ROC mide la capacidad del modelo para distinguir entre las clases positiva y negativa. Un valor de 0.776 indica que el modelo tiene un rendimiento decente en términos de separación de clases, pero no es excelente.

Aunque el AUC-ROC es relativamente bueno, el F1 Score bajo sugiere que el modelo no está equilibrando bien la precisión y el recall.

<div class="alert alert-block alert-success">
<b>Comentario del revisor (1ra Iteración)</b> <a class=“tocSkip”></a>

Excelente trabajo implemetando el modelo de regresión lógistica sin ajuste de equilibrio de clases, esto permite tener una base para ver el efecto de las técnicas empleadas

<div class="alert alert-block alert-warning">
<b>Comentario del revisor (1ra Iteración)</b> <a class=“tocSkip”></a>

Al importar librerías en un notebook de Python, una buena práctica es incluir en la primera celda todas las que vas a emplear en el proyecto, organizándolas por orden de importancia o categorías: primero las librerías estándar, seguidas de las de terceros y, finalmente, cualquier módulo personalizado. Además, documenta librerías menos comunes con comentarios para facilitar su comprensión

</div>
</div>

# Paso 3.  Mejoras al modelo

# Modelo de Regresión logistica

# Submuestreo

In [6]:
#SUBMUESTREO

from sklearn.utils import shuffle

def downsample(X, y, fraction):
    # Asegurarse de que X e y tengan los mismos índices
    X = pd.DataFrame(X).reset_index(drop=True)
    y = pd.Series(y).reset_index(drop=True)
    
    # Filtrar las clases
    X_zeros = X[y == 0]
    X_ones = X[y == 1]
    y_zeros = y[y == 0]
    y_ones = y[y == 1]

    # Submuestrear la clase mayoritaria
    X_zeros_downsampled = X_zeros.sample(frac=fraction, random_state=12345)
    y_zeros_downsampled = y_zeros.sample(frac=fraction, random_state=12345)

    # Combinar las clases
    X_downsampled = pd.concat([X_zeros_downsampled, X_ones])
    y_downsampled = pd.concat([y_zeros_downsampled, y_ones])

    # Mezclar los datos
    return shuffle(X_downsampled, y_downsampled, random_state=12345)

# Aplicar submuestreo
X_train_downsampled, y_train_downsampled = downsample(X_train, y_train, fraction=0.1)

# Verificar el tamaño de los datos submuestreados
print("Tamaño de X_train_downsampled:", X_train_downsampled.shape)
print("Tamaño de y_train_downsampled:", y_train_downsampled.shape)

Tamaño de X_train_downsampled: (1970, 11)
Tamaño de y_train_downsampled: (1970,)


In [7]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score, roc_auc_score

# Entrenar el modelo con datos submuestreados
model_downsampled = LogisticRegression(random_state=12345, solver='liblinear')
model_downsampled.fit(X_train_downsampled, y_train_downsampled)

# Predecir en el conjunto de validación
y_pred_downsampled = model_downsampled.predict(X_val)

# Calcular F1 y AUC-ROC
f1_downsampled = f1_score(y_val, y_pred_downsampled)
auc_roc_downsampled = roc_auc_score(y_val, model_downsampled.predict_proba(X_val)[:, 1])

print(f'F1 Score (Submuestreo): {f1_downsampled:.3f}')
print(f'AUC-ROC (Submuestreo): {auc_roc_downsampled:.3f}')

F1 Score (Submuestreo): 0.412
AUC-ROC (Submuestreo): 0.777


Con el método de submuestreo se observa un mejor rendimiento del modelo, una mejora en el F1 Score (0.41) y el AUC-ROC se mantuvo en 0.77. 

# Sobremuestreo

In [10]:
import pandas as pd
import numpy as np
from sklearn.utils import shuffle
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score, roc_auc_score

# Función para sobremuestrear la clase minoritaria
def oversample(X, y):
    # Asegurarse de que X e y tengan los mismos índices
    X = pd.DataFrame(X).reset_index(drop=True)
    y = pd.Series(y).reset_index(drop=True)
    
    # Filtrar las clases
    X_zeros = X[y == 0]
    X_ones = X[y == 1]
    y_zeros = y[y == 0]
    y_ones = y[y == 1]

    # Calcular cuántas veces se debe duplicar la clase minoritaria
    n_zeros = len(X_zeros)
    n_ones = len(X_ones)
    repeat_times = n_zeros // n_ones  # Número de veces que se duplicará la clase minoritaria

    # Duplicar la clase minoritaria
    X_ones_oversampled = pd.concat([X_ones] * repeat_times)
    y_ones_oversampled = pd.concat([y_ones] * repeat_times)

    # Combinar las clases
    X_oversampled = pd.concat([X_zeros, X_ones_oversampled])
    y_oversampled = pd.concat([y_zeros, y_ones_oversampled])

    # Mezclar los datos
    return shuffle(X_oversampled, y_oversampled, random_state=12345)

# Aplicar sobremuestreo
X_train_oversampled, y_train_oversampled = oversample(X_train, y_train)

# Verificar el nuevo tamaño de los datos
print("Distribución de clases después de sobremuestreo:")
print(pd.Series(y_train_oversampled).value_counts())

# Entrenar el modelo con datos sobremuestreados
model_oversampled = LogisticRegression(random_state=12345, solver='liblinear')
model_oversampled.fit(X_train_oversampled, y_train_oversampled)

# Predecir en el conjunto de validación
y_pred_oversampled = model_oversampled.predict(X_val)

# Calcular F1 y AUC-ROC
f1_oversampled = f1_score(y_val, y_pred_oversampled)
auc_roc_oversampled = roc_auc_score(y_val, model_oversampled.predict_proba(X_val)[:, 1])

print(f'F1 Score (Sobremuestreo): {f1_oversampled:.3f}')
print(f'AUC-ROC (Sobremuestreo): {auc_roc_oversampled:.3f}')

Distribución de clases después de sobremuestreo:
0    5589
1    4233
Name: Exited, dtype: int64
F1 Score (Sobremuestreo): 0.512
AUC-ROC (Sobremuestreo): 0.779


Distribución de Clases: El sobremuestreo ha aumentado el número de instancias de la clase minoritaria (1) de 1,411 a 4,233.

Aunque no se ha alcanzado un equilibrio perfecto (5,589 vs. 4,233), la clase minoritaria está mejor representada.

F1 Score (0.512): El F1 Score ha mejorado en comparación con el modelo básico (0.319), lo que indica que el sobremuestreo ha ayudado a equilibrar mejor la precisión y el recall.

Sin embargo, el valor aún está por debajo del objetivo de 0.59.

AUC-ROC (0.779): El AUC-ROC se mantiene relativamente alto, lo que sugiere que el modelo tiene una buena capacidad para distinguir entre las clases.

# Ponderación de clases

In [11]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score, roc_auc_score

# Cargar los datos
data = pd.read_csv('/datasets/Churn.csv')

# Preprocesamiento
data = data.drop(['RowNumber', 'CustomerId', 'Surname'], axis=1)  # Eliminar columnas irrelevantes
data = pd.get_dummies(data, columns=['Geography', 'Gender'], drop_first=True)  # Codificar variables categóricas

# Imputar valores nulos en 'Tenure'
median_tenure = data['Tenure'].median()
data['Tenure'].fillna(median_tenure, inplace=True)

# Separar características (X) y objetivo (y)
X = data.drop('Exited', axis=1)
y = data['Exited']

# Dividir en conjuntos de entrenamiento, validación y prueba
X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size=0.3, random_state=12345)
X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=12345)

# Estandarizar características numéricas
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_val = scaler.transform(X_val)
X_test = scaler.transform(X_test)

# Entrenar el modelo con ponderación de clases
model_weighted = LogisticRegression(random_state=12345, solver='liblinear', class_weight='balanced')
model_weighted.fit(X_train, y_train)

# Predecir en el conjunto de validación
y_pred_weighted = model_weighted.predict(X_val)

# Calcular F1 y AUC-ROC
f1_weighted = f1_score(y_val, y_pred_weighted)
auc_roc_weighted = roc_auc_score(y_val, model_weighted.predict_proba(X_val)[:, 1])

print(f'F1 Score (Ponderación): {f1_weighted:.3f}')
print(f'AUC-ROC (Ponderación): {auc_roc_weighted:.3f}')

F1 Score (Ponderación): 0.504
AUC-ROC (Ponderación): 0.780


# Ajuste de hiperparametros

In [12]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score, roc_auc_score

# Definir los hiperparámetros a ajustar
param_grid = {
    'C': [0.01, 0.1, 1, 10],  # Parámetro de regularización
    'solver': ['liblinear', 'lbfgs'],  # Algoritmo de optimización
    'class_weight': [None, 'balanced']  # Ponderación de clases
}

# Crear el modelo base
model = LogisticRegression(random_state=12345)

# Configurar GridSearchCV
grid_search = GridSearchCV(
    estimator=model,
    param_grid=param_grid,
    scoring='f1',  # Métrica a optimizar (F1 Score)
    cv=3,  # Número de divisiones en la validación cruzada
    n_jobs=-1  # Usar todos los núcleos del procesador
)

# Ajustar GridSearchCV a los datos de entrenamiento
grid_search.fit(X_train, y_train)

# Mejores hiperparámetros encontrados
print("Mejores hiperparámetros:", grid_search.best_params_)

# Predecir en el conjunto de validación con el mejor modelo
y_pred_grid = grid_search.predict(X_val)

# Calcular F1 y AUC-ROC
f1_grid = f1_score(y_val, y_pred_grid)
auc_roc_grid = roc_auc_score(y_val, grid_search.predict_proba(X_val)[:, 1])

print(f'F1 Score (GridSearchCV): {f1_grid:.3f}')
print(f'AUC-ROC (GridSearchCV): {auc_roc_grid:.3f}')

Mejores hiperparámetros: {'C': 0.01, 'class_weight': 'balanced', 'solver': 'lbfgs'}
F1 Score (GridSearchCV): 0.503
AUC-ROC (GridSearchCV): 0.780


In [13]:
#Sobremuestreo + Ajuste hiperparametros

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score, roc_auc_score
from sklearn.utils import shuffle

# Función para sobremuestrear la clase minoritaria
def oversample(X, y):
    X = pd.DataFrame(X).reset_index(drop=True)
    y = pd.Series(y).reset_index(drop=True)
    
    X_zeros = X[y == 0]
    X_ones = X[y == 1]
    y_zeros = y[y == 0]
    y_ones = y[y == 1]

    n_zeros = len(X_zeros)
    n_ones = len(X_ones)
    repeat_times = n_zeros // n_ones  # Número de veces que se duplicará la clase minoritaria

    X_ones_oversampled = pd.concat([X_ones] * repeat_times)
    y_ones_oversampled = pd.concat([y_ones] * repeat_times)

    X_oversampled = pd.concat([X_zeros, X_ones_oversampled])
    y_oversampled = pd.concat([y_zeros, y_ones_oversampled])

    return shuffle(X_oversampled, y_oversampled, random_state=12345)

# Cargar los datos
data = pd.read_csv('/datasets/Churn.csv')

# Preprocesamiento
data = data.drop(['RowNumber', 'CustomerId', 'Surname'], axis=1)  # Eliminar columnas irrelevantes
data = pd.get_dummies(data, columns=['Geography', 'Gender'], drop_first=True)  # Codificar variables categóricas

# Imputar valores nulos en 'Tenure'
median_tenure = data['Tenure'].median()
data['Tenure'].fillna(median_tenure, inplace=True)

# Separar características (X) y objetivo (y)
X = data.drop('Exited', axis=1)
y = data['Exited']

# Dividir en conjuntos de entrenamiento, validación y prueba
X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size=0.3, random_state=12345)
X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=12345)

# Estandarizar características numéricas
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_val = scaler.transform(X_val)
X_test = scaler.transform(X_test)

# Aplicar sobremuestreo
X_train_oversampled, y_train_oversampled = oversample(X_train, y_train)

# Definir los hiperparámetros a ajustar
param_grid = {
    'C': [0.01, 0.1, 1, 10],  # Parámetro de regularización
    'solver': ['liblinear', 'lbfgs'],  # Algoritmo de optimización
    'class_weight': [None, 'balanced']  # Ponderación de clases
}

# Crear el modelo base
model = LogisticRegression(random_state=12345)

# Configurar GridSearchCV
grid_search = GridSearchCV(
    estimator=model,
    param_grid=param_grid,
    scoring='f1',  # Métrica a optimizar (F1 Score)
    cv=3,  # Número de divisiones en la validación cruzada
    n_jobs=-1  # Usar todos los núcleos del procesador
)

# Ajustar GridSearchCV a los datos sobremuestreados
grid_search.fit(X_train_oversampled, y_train_oversampled)

# Mejores hiperparámetros encontrados
print("Mejores hiperparámetros:", grid_search.best_params_)

# Predecir en el conjunto de validación con el mejor modelo
y_pred_grid = grid_search.predict(X_val)

# Calcular F1 y AUC-ROC
f1_grid = f1_score(y_val, y_pred_grid)
auc_roc_grid = roc_auc_score(y_val, grid_search.predict_proba(X_val)[:, 1])

print(f'F1 Score (Sobremuestreo + GridSearchCV): {f1_grid:.3f}')
print(f'AUC-ROC (Sobremuestreo + GridSearchCV): {auc_roc_grid:.3f}')

Mejores hiperparámetros: {'C': 0.01, 'class_weight': 'balanced', 'solver': 'liblinear'}
F1 Score (Sobremuestreo + GridSearchCV): 0.500
AUC-ROC (Sobremuestreo + GridSearchCV): 0.780


Se procede a probar otros tipos de modelos para acercarse a un mejor F1 Score, de minimo 59%

# Modelo arbol de decisión

In [14]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import f1_score, roc_auc_score

# Definir los hiperparámetros a ajustar
param_grid_tree = {
    'max_depth': [3, 5, 7, 10, None],  # Profundidad máxima del árbol
    'min_samples_split': [2, 5, 10],  # Mínimo número de muestras para dividir un nodo
    'min_samples_leaf': [1, 2, 4],  # Mínimo número de muestras en una hoja
    'class_weight': [None, 'balanced']  # Ponderación de clases
}

# Crear el modelo base
tree_model = DecisionTreeClassifier(random_state=12345)

# Configurar GridSearchCV
grid_search_tree = GridSearchCV(
    estimator=tree_model,
    param_grid=param_grid_tree,
    scoring='f1',  # Métrica a optimizar (F1 Score)
    cv=3,  # Número de divisiones en la validación cruzada
    n_jobs=-1  # Usar todos los núcleos del procesador
)

# Ajustar GridSearchCV a los datos sobremuestreados
grid_search_tree.fit(X_train_oversampled, y_train_oversampled)

# Mejores hiperparámetros encontrados
print("Mejores hiperparámetros (Árbol de Decisión):", grid_search_tree.best_params_)

# Predecir en el conjunto de validación con el mejor modelo
y_pred_tree = grid_search_tree.predict(X_val)

# Calcular F1 y AUC-ROC
f1_tree = f1_score(y_val, y_pred_tree)
auc_roc_tree = roc_auc_score(y_val, grid_search_tree.predict_proba(X_val)[:, 1])

print(f'F1 Score (Árbol de Decisión): {f1_tree:.3f}')
print(f'AUC-ROC (Árbol de Decisión): {auc_roc_tree:.3f}')

Mejores hiperparámetros (Árbol de Decisión): {'class_weight': 'balanced', 'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2}
F1 Score (Árbol de Decisión): 0.486
AUC-ROC (Árbol de Decisión): 0.676


# Modelo Bosque aleatorio

# Con ajuste de hiperparametros y sobremuestreo

In [15]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import f1_score, roc_auc_score

# Definir los hiperparámetros a ajustar
param_grid_rf = {
    'n_estimators': [50, 100, 200],  # Número de árboles en el bosque
    'max_depth': [3, 5, 7, 10, None],  # Profundidad máxima de cada árbol
    'min_samples_split': [2, 5, 10],  # Mínimo número de muestras para dividir un nodo
    'min_samples_leaf': [1, 2, 4],  # Mínimo número de muestras en una hoja
    'class_weight': [None, 'balanced']  # Ponderación de clases
}

# Crear el modelo base
rf_model = RandomForestClassifier(random_state=12345)

# Configurar GridSearchCV
grid_search_rf = GridSearchCV(
    estimator=rf_model,
    param_grid=param_grid_rf,
    scoring='f1',  # Métrica a optimizar (F1 Score)
    cv=3,  # Número de divisiones en la validación cruzada
    n_jobs=-1  # Usar todos los núcleos del procesador
)

# Ajustar GridSearchCV a los datos sobremuestreados
grid_search_rf.fit(X_train_oversampled, y_train_oversampled)

# Mejores hiperparámetros encontrados
print("Mejores hiperparámetros (Bosque Aleatorio):", grid_search_rf.best_params_)

# Predecir en el conjunto de validación con el mejor modelo
y_pred_rf = grid_search_rf.predict(X_val)

# Calcular F1 y AUC-ROC
f1_rf = f1_score(y_val, y_pred_rf)
auc_roc_rf = roc_auc_score(y_val, grid_search_rf.predict_proba(X_val)[:, 1])

print(f'F1 Score (Bosque Aleatorio): {f1_rf:.3f}')
print(f'AUC-ROC (Bosque Aleatorio): {auc_roc_rf:.3f}')

Mejores hiperparámetros (Bosque Aleatorio): {'class_weight': 'balanced', 'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 200}
F1 Score (Bosque Aleatorio): 0.628
AUC-ROC (Bosque Aleatorio): 0.850


El modelo de Bosque Aleatorio es el que mejores resultados presenta con ajustes de sobremuestreo y en hiperparametros, con un F1 Score de 62.8%, superando el mínimo de 59%.También se encuentra una mejora en el AUC-ROC con un 85%.

# Con submuestreo

In [22]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import f1_score, roc_auc_score
from sklearn.utils import shuffle

# Función para submuestrear la clase mayoritaria
def downsample(X, y, fraction):
    X = pd.DataFrame(X).reset_index(drop=True)
    y = pd.Series(y).reset_index(drop=True)
    
    # Filtrar las clases
    X_zeros = X[y == 0]
    X_ones = X[y == 1]
    y_zeros = y[y == 0]
    y_ones = y[y == 1]

    # Submuestrear la clase mayoritaria
    X_zeros_downsampled = X_zeros.sample(frac=fraction, random_state=12345)
    y_zeros_downsampled = y_zeros.sample(frac=fraction, random_state=12345)

    # Combinar las clases
    X_downsampled = pd.concat([X_zeros_downsampled, X_ones])
    y_downsampled = pd.concat([y_zeros_downsampled, y_ones])

    # Mezclar los datos
    return shuffle(X_downsampled, y_downsampled, random_state=12345)

# Aplicar submuestreo
fraction = 0.5  # Fracción de la clase mayoritaria a mantener
X_train_downsampled, y_train_downsampled = downsample(X_train, y_train, fraction)

# Verificar el nuevo tamaño de los datos
print("Distribución de clases después de submuestreo:")
print(pd.Series(y_train_downsampled).value_counts())

# Entrenar el modelo de Bosque Aleatorio sin ajuste de hiperparámetros
rf_model = RandomForestClassifier(random_state=12345, class_weight='balanced')
rf_model.fit(X_train_downsampled, y_train_downsampled)

# Predecir en el conjunto de validación
y_pred_rf = rf_model.predict(X_val)

# Calcular F1 y AUC-ROC
f1_rf = f1_score(y_val, y_pred_rf)
auc_roc_rf = roc_auc_score(y_val, rf_model.predict_proba(X_val)[:, 1])

print(f'F1 Score (Bosque Aleatorio - Submuestreo): {f1_rf:.3f}')
print(f'AUC-ROC (Bosque Aleatorio - Submuestreo): {auc_roc_rf:.3f}')



Distribución de clases después de submuestreo:
0    2794
1    1411
Name: Exited, dtype: int64
F1 Score (Bosque Aleatorio - Submuestreo): 0.632
AUC-ROC (Bosque Aleatorio - Submuestreo): 0.854


El modelo de Bosque Aleatorio es el que mejores resultados presenta con ajustes de submuestreo, con un F1 Score de 63.2%, superando el mínimo de 59%.También se encuentra una mejora en el AUC-ROC con un 85%.

# Paso 4 - Prueba final

In [24]:
# Entrenar el modelo de Bosque Aleatorio sin ajuste de hiperparámetros
rf_model = RandomForestClassifier(random_state=12345, class_weight='balanced')
rf_model.fit(X_train_downsampled, y_train_downsampled)

# Obtener las probabilidades de la clase positiva (1) en el conjunto de validación
y_probs_val = rf_model.predict_proba(X_val)[:, 1]

# Encontrar el umbral óptimo que maximiza el F1 Score
precision, recall, thresholds = precision_recall_curve(y_val, y_probs_val)
f1_scores = 2 * (precision * recall) / (precision + recall)
optimal_threshold = thresholds[np.argmax(f1_scores)]

print(f'Umbral óptimo: {optimal_threshold:.3f}')

# Aplicar el umbral óptimo en el conjunto de validación
y_pred_val_optimal = (y_probs_val >= optimal_threshold).astype(int)

# Calcular F1 y AUC-ROC en el conjunto de validación con el umbral óptimo
f1_val_optimal = f1_score(y_val, y_pred_val_optimal)
auc_roc_val_optimal = roc_auc_score(y_val, y_probs_val)

print(f'F1 Score (Validación - Umbral Óptimo): {f1_val_optimal:.3f}')
print(f'AUC-ROC (Validación - Umbral Óptimo): {auc_roc_val_optimal:.3f}')

# Aplicar el umbral óptimo en el conjunto de prueba
y_probs_test = rf_model.predict_proba(X_test)[:, 1]
y_pred_test_optimal = (y_probs_test >= optimal_threshold).astype(int)

# Calcular F1 y AUC-ROC en el conjunto de prueba con el umbral óptimo
f1_test_optimal = f1_score(y_test, y_pred_test_optimal)
auc_roc_test_optimal = roc_auc_score(y_test, y_probs_test)

print(f'F1 Score (Prueba Final - Umbral Óptimo): {f1_test_optimal:.3f}')
print(f'AUC-ROC (Prueba Final - Umbral Óptimo): {auc_roc_test_optimal:.3f}')

Umbral óptimo: 0.530
F1 Score (Validación - Umbral Óptimo): 0.637
AUC-ROC (Validación - Umbral Óptimo): 0.854
F1 Score (Prueba Final - Umbral Óptimo): 0.593
AUC-ROC (Prueba Final - Umbral Óptimo): 0.853


<div class="alert alert-block alert-success">
<b>Comentario del revisor (1ra Iteración)</b> <a class=“tocSkip”></a>

Excelente trabajo implementado las técnicas para el manejo de datos desbalanceados. Has logrado mejorar el rendimiento de los modelos de clasificación que planteaste, Aunque es importante saber que los efectos en el rendimiento dependerá de los modelos utilizados, las técnicas de equilibrio empleadas y de la estructura misma de los datos. En otras circuntancias, los resultados pueden variar, por eso es clave hacer varias pruebas con el fin de identificar la mejor combinación o estrategía que nos de el mejor resultado

</div>

# Conclusión

Luego de aplicar diversas mejoras (sobremuestreo, el submuestreo, la ponderación de clases, ajuste de umbral de clasificación) y a diferentes modelos (regresión logistica, arbol de decisión y bosque), se puede concluir que el modelo de Bosque Aleatorio con submuestreo y el umbral ajustado ha demostrado un rendimiento excelente, alcanzando un F1 Score de 59,3% y un AUC-ROC de 85,3% en el conjunto de prueba.

<div class="alert alert-block alert-success">
<b>Comentario del revisor (1ra Iteración)</b> <a class=“tocSkip”></a>

Tu enfoque en la optimización de modelos mediante técnicas como el sobremuestreo, submuestreo y ajuste de umbrales demuestra un manejo sólido de estrategias para abordar problemas de desbalanceo de clases. Es destacable cómo has evaluado múltiples algoritmos, identificando que el Bosque Aleatorio, combinado con submuestreo y un umbral ajustado, ofrece el mejor rendimiento. La elección de métricas como el F1 Score y el AUC-ROC refleja una comprensión profunda de la importancia de equilibrar precisión y recall en contextos desbalanceados. Para futuros trabajos, podrías explorar la interpretabilidad del modelo, utilizando técnicas como la importancia de características o SHAP values, lo que podría proporcionar insights adicionales sobre los factores que más influyen en las predicciones

</div>